crates/base/src/global/veci8.rs

```diff
@@ -0,0 +1,230 @@
+use crate::{
+    global::Veci8Owned,
+    scalar::{F32, I8},
+};
+
+use super::Veci8Borrowed;
+
+pub fn dot(x: &[I8], y: &[I8]) -> F32 {
+    #[cfg(target_arch = "x86_64")]
+    {
+        if detect::x86_64::test_avx512vnni() {
+            return unsafe { dot_i8_avx512vnni(x, y) };
+        }
+    }
+    dot_i8_fallback(x, y)
+}
+
+#[multiversion::multiversion(targets(
+    "x86_64/x86-64-v4",
+    "x86_64/x86-64-v3",
+    "x86_64/x86-64-v2",
+    "aarch64+neon"
+))]
+fn dot_i8_fallback(x: &[I8], y: &[I8]) -> F32 {
+    // i8 * i8 fall in range of i16. Since our length is less than (2^16 - 1), the result won't overflow.
+    let mut sum = 0;
+    assert_eq!(x.len(), y.len());
+    let length = x.len();
+    // according to https://godbolt.org/z/ff48vW4es, this loop will be autovectorized
+    for i in 0..length {
+        sum += (x[i].0 as i16 * y[i].0 as i16) as i32;
+    }
+    F32(sum as f32)
+}
+
+#[cfg(target_arch = "x86_64")]
+#[target_feature(enable = "avx512f,avx512bw,avx512vnni,bmi2")]
+unsafe fn dot_i8_avx512vnni(x: &[I8], y: &[I8]) -> F32 {
+    use std::arch::x86_64::*;
+
+    assert_eq!(x.len(), y.len());
+    let mut sum = 0;
+    let mut i = x.len();
+    let mut p_x = x.as_ptr() as *const i8;
+    let mut p_y = y.as_ptr() as *const i8;
+    let mut vec_x;
+    let mut vec_y;
+    unsafe {
+        let mut result = _mm512_setzero_si512();
+        let zero = _mm512_setzero_si512();
+        while i > 0 {
+            if i < 64 {
+                let mask = _bzhi_u64(0xFFFF_FFFF_FFFF_FFFF, i as u32);
+                vec_x = _mm512_maskz_loadu_epi8(mask, p_x);
+                vec_y = _mm512_maskz_loadu_epi8(mask, p_y);
+                i = 0;
+            } else {
+                vec_x = _mm512_loadu_epi8(p_x);
+                vec_y = _mm512_loadu_epi8(p_y);
+                i -= 64;
+                p_x = p_x.add(64);
+                p_y = p_y.add(64);
+            }
+            // there are only _mm512_dpbusd_epi32 support, dpbusd will zeroextend a[i] and signextend b[i] first, so we need to convert a[i] positive and change corresponding b[i] to get right result.
+            // And because we use -b[i] here, the range of quantization should be [-127, 127] instead of [-128, 127] to avoid overflow.
```

Comment:
We didn't check `i8` values are in `[-127,+127]` for users' input so it may lead to problems. Saturate arithmetics could help here.

Reply:
Great catch. Thanks.